### PR TITLE
ui: upgrade React 15.4 => 16.3

### LINKED ISSUE #18795: ui: Upgrade to React 16 [460]
React 16 has been released, and along with it a number of purported performance benefits.  We should allow some reasonable time for stability (and for third-party libraries to catch up), then investigate moving over.  React 16 has a very small number of [breaking changes](https://facebook.github.io/react/blog/2017/09/26/react-v16.0.html#breaking-changes), but I don't believe any of them apply to us; once the third-party ecosystem is caught up, we should be able to upgrade without issue.

### PR DESCRIPTION
**Pull request by vilterp**
Fixes #18795

Release note: None

### CODE REVIEW

File: pkg/ui/package.json [hunk 1]
```diff
@@ -7,19 +7,17 @@
...
-    "codemirror": "^5.25.2",
     "combokeys": "^2.4.6",
     "highlight.js": "^9.10.0",
     "lodash": "^4.17.4",
     "moment": "^2.18.1",
     "nvd3": "^1.8.5",
     "raw-loader": "^0.5.1",
     "rc-progress": "^2.2.1",
-    "react": "^15.4.2",
+    "react": "^16.3.1",
```
Comment (couchand) [inline pkg/ui/package.json]: plz also update the `@types` for these

File: pkg/ui/src/views/app/containers/timewindow/timewindow.spec.tsx [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
+import Enzyme, { shallow } from "enzyme";
+import Adapter from "enzyme-adapter-react-16";
 import * as sinon from "sinon";
 import moment from "moment";
 import _ from "lodash";
 
 import { TimeWindowManagerUnconnected as TimeWindowManager } from "./";
 import * as timewindow from "src/redux/timewindow";
 
+Enzyme.configure({ adapter: new Adapter() });
```
Comment (couchand) [inline pkg/ui/src/views/app/containers/timewindow/timewindow.spec.tsx]: maybe put this config in a test helper?

**Review by couchand — CHANGES_REQUESTED** [review]
kind of surprised this is all that's needed, but great!

**Comment by couchand** [comment]
oh goodness, and a bunch of types incompatibilities 😱

**Comment by couchand** [comment]
You're probably not seeing type issues cause `@types/react` and `@types/react-dom` are still out-of-date...

**Comment by couchand** [comment]
Re: `react-sticky`, already have an open issue to get rid of it, I don't think we need it at all.  #19522

**Comment by vilterp** [comment]
Ok, after a day of wrangling, I think I've gotten most of the `@types/` upgrades, and we're still at one peer dep issue (`react-sticky`); may attempt to rip it out per #19522.

**Comment by mrtracy** [comment]
The change to Table specializations is LGTM

**Comment by couchand** [comment]
When you squash can you also link the commit to #18795 please?

File: pkg/ui/package.json (hunk 1 above)
Comment (vilterp) [reply pkg/ui/package.json]: Done; think I got them all.

File: pkg/ui/src/views/app/containers/layout/index.tsx [hunk 3]
```diff
@@ -15,14 +14,16 @@ import AlertBanner from "src/views/app/containers/alertBanner";
...
-    return <div>
-      <Helmet titleTemplate="%s | Cockroach Console" defaultTitle="Cockroach Console" />
-      <TimeWindowManager/>
-      <AlertBanner/>
-      <NavigationBar/>
-      <StickyContainer className="page">
-        { this.props.children }
-      </StickyContainer>
-    </div>;
+    return (
```
Comment (couchand) [inline pkg/ui/src/views/app/containers/layout/index.tsx]: +1 noticed this when adding the titles, meant to come back and do the reformat but forgot.  thanks!

**Comment by couchand** [comment]
looks great!  request from @josueeee: can you add some `padding: 12 0` to the config

**Review by couchand — APPROVED** [review]

**Comment by vilterp** [comment]
Problems with `position: sticky`:
- thin grey line shows up under header whether we are scrolled down or not

Problems with `react-sticky`:
- current version has peer deps issue (depends on React 15)
- new version doesn't have typings

**Comment by vilterp** [comment]
fixed nit

bors r+

**Pull request merged**